earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +151,173 @@ class khMTSet
     khLockGuard guard(mutex);
     return std::vector<T>(set.begin(), set.end());
   }
+  void clear(void) {
+    khLockGuard guard(mutex);
+    set.clear();
+  }
+  size_t size(void) {
+    khLockGuard guard(mutex);
+    return set.size();
+  }
 };
 
 
+template <class K, class V>
+class khMTMap
+{
+ protected:
+  typedef std::unordered_map<K,V> Map;
+  Map     map;
+  mutable std::mutex mutex;
+
+ private:
+  // private and unimplimented
+  khMTMap& operator=(const khMTMap&);
+
+ public:
+  khMTMap(void) { }
+
+  void insert(const std::pair<K,V> &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    (void)map.insert(val);
+  }
+  void erase(const K &val) {
+    std::lock_guard<std::mutex> guard(mutex);
+    if (map.find(val) != map.end()) {
+      (void)map.erase(val);
+    }
+  }
+  size_t size(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.size();
+  }
+  V& operator[](const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map[key];
+  }
+  void clear(void) {
+    std::lock_guard<std::mutex> guard(mutex);
+    map.clear();
+  }
+  std::vector<K> keys(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    std::vector<K> thekeys;
+    thekeys.reserve(map.size());
+    for (const auto& p : map) {
+      thekeys.push_back(p.first);
+    }
+    return thekeys;
+  }
+  bool contains(const K& key) {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.find(key) != map.end();
+  }
+  bool empty(void) const {
+    std::lock_guard<std::mutex> guard(mutex);
+    return map.empty();
+  }
+  khMTMap(const khMTMap&) {
+    // todo: this is only used for setting up an empty dirtymap by AssetD and AssetVersionD
+    // should probably really remove this
+  }
+};
+
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but inheriting from vector publicly isn't great
+template <typename T>
+class MTVector : public std::vector<T> {
+  private:
+    mutable std::mutex mtx;
+
+    MTVector(const MTVector& a, const std::lock_guard<std::mutex> &) : Base(a) { }
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+    void push_back(const T& v) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::push_back(v);
+    }
+    MTVector() : Base() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, std::lock_guard<std::mutex>(a.mtx)) { }
+    const T& operator[] (const size_t idx) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::operator[](idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return true;
```

Comment:
Is this class intended to be permanent? There's several unimplemented functions, like this one that says all MTVectors are the same.

Reply:
oops.  No this shouldn't be here but I forgot about it while fixing other problems.  I will fix this to do a proper comparison.